### DISCUSSION TITLE
Expand keptn CLI trigger * to trigger any sequence?

### DISCUSSION THREAD

**Discussion by ghost** [keptn/keptn#4839] [Ideas]
Is it a good idea to expand the keptn CLI to trigger **any** sequence, not only `delivery` and `evaluation`

**Comment by ghost** [comment]
Accepted and implementation happening under: https://github.com/keptn/keptn/issues/6060

**Comment by afzal442** [comment]
Regarding the issue #6060 

> Do you need any assistance from the Keptn team?

Thanks so much @Dynatrace-Adam-Gardner for offering that help. 🪖  And yes, I apologise for not being involved in it for a longer period of time. I truly need assistance in setting up Keptn CLI and its dependencies from the Keptn team. Let me know if we can meet each other. 🏷️ 
Although I went through the [guide](https://tutorials.keptn.sh/?cat=quickstart), I'm confused here about how to set up CLI and test its commands locally using any sample application or any use case.
Thanks! 😄 

// @odubajDT

**Reply by odubajDT** [reply]
Hi @afzal442 if you need assistance, I am happy to help :) Do you have Keptn already installed ? If no, you can use this documentation steps to install and run it on your local K3d machine. Then you can also use and check your changes against integration tests. Please be aware, that you need to have the latest Keptn-0.12.0 installed. If yes, please clone this repository https://github.com/keptn/keptn and in the `/cli` directory, there you have code of the CLI command. It's implemented in Go, so you need to have it installed. If you have it, just type `go build` from the `/cli` directory and the binary will be build, then you can use the binary to check and run you local changes. Please be aware, intead of using the installed `keptn` command, you will use your build binary, which will be called `cli` (or whatever will be created in your local `/cli` directory), so you can execute it by calling `./cli <commands>`. Just type `./cli --help` for more info. If any questions, just reach out. :)

**Reply by afzal442** [reply]
That's really sharp made out. I'm retreating over it. I have my gke-cluster running up against this project. Can I build binary files for that and test the commands as you specified on top of that? Thanks.
// @odubajDT

**Reply by odubajDT** [reply]
Yes, gke-cluster is also an option :)

**Reply by afzal442** [reply]
hi @odubajDT, I have been doing some exploration here. I tried running `keptn trigger delivery sequence=my-sequence --project=sockshop --stage=dev --service=carts --image=nginx:1.14`
I fell into some trouble here to figure it out
`Availability of provided image cannot be checked.`
Let me know how I can check  Continuous Delivery on top of use cases like mentioned in the issue using keptn CLI tools.
Thanks

**Reply by odubajDT** [reply]
Hi @afzal442, you actually need to split the docker image name and tag, like in the following example:

`keptn trigger delivery --project=sockshop --service=carts --stage=dev --image=nginx --tag=1.14 --sequence=delivery`

**Reply by ghost** [reply]
@afzal442 to implement the `keptn trigger sequence` command as per this discussion, have a look at [trigger_delivery.go](https://github.com/keptn/keptn/blob/master/cli/cmd/trigger_delivery.go) as a template.

**Reply by ghost** [reply]
@afzal442 here's a version I got working. Note it has **zero** error checking, no tests and I'm a Go beginner so I know there will be improvements, but it works:
```
keptn trigger sequence --sequence=hello --project=hello-world --service=tasks --stage=dev
```

https://gist.github.com/Dynatrace-Adam-Gardner/269c08455a05b5e8cbb083bbe4532d7c

<img width="728" alt="image" src="https://user-images.githubusercontent.com/13639658/151931929-a580cb33-8b66-4aed-9a57-a370480907c8.png">

**Reply by afzal442** [reply]
@Dynatrace-Adam-Gardner that's awesome! 😄 Basically once you know the development setup and configure it, you get your job more than half done. 🏷️

**Reply by afzal442** [reply]
> have a look at trigger_delivery.go as a template.

Thanks @Dynatrace-Adam-Gardner 
That's exactly what I have been looking at. I have my that work on my local machine. I'm trying to run and test.
// @odubajDT

**Comment by afzal442** [comment]
@odubajDT I have issue with `keptn auth`. I have gke cluster up and keptn deployed to it. 
`kubectl -n keptn get service api-gateway-nginx`
getting this err! `Error from server (NotFound): services "api-gateway-nginx" not found`
As I told I have been working at that using that trigger_delivery.go, I also created a gist for that. Thanks
https://gist.github.com/afzal442/15971778a6b6928cc36be0450c7e84c9

**Reply by afzal442** [reply]
`./cli trigger sequence --sequence=delivery --project=sockshop --service=carts --stage=dev`

**Reply by odubajDT** [reply]
can you please execute `kubectl -n keptn get svc` and paste the result here ?

**Reply by afzal442** [reply]
sorry @odubajDT I upgraded keptn using helm chart, but it's now having another issue. I tried to access Keptn via a Port-forward
`keptn auth --endpoint=$KEPTN_ENDPOINT --api-token=$KEPTN_API_TOKEN`

`Kube context has been changed to gke_xxxx_asia-south1-a_cluster-2
Starting to authenticate
Authentication was unsuccessful. invalid character '<' looking for beginning of value`
FYI,
`kubectl -n keptn get svc`
```
NAME                    TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)                                                 AGE
api-gateway-nginx       ClusterIP   10.68.1.160    <none>        80/TCP                                                  12m
api-service             ClusterIP   10.68.11.210   <none>        8080/TCP                                                12m
approval-service        ClusterIP   10.68.15.239   <none>        8080/TCP                                                12m
bridge                  ClusterIP   10.68.11.128   <none>        8080/TCP                                                12m
configuration-service   ClusterIP   10.68.8.74     <none>        8080/TCP                                                12m
keptn-mongo             ClusterIP   10.68.7.207    <none>        27017/TCP                                               12m
keptn-nats-cluster      ClusterIP   None           <none>        4222/TCP,6222/TCP,8222/TCP,7777/TCP,7422/TCP,7522/TCP   12m
lighthouse-service      ClusterIP   10.68.3.160    <none>        8080/TCP                                                12m
mongodb-datastore       ClusterIP   10.68.7.182    <none>        8080/TCP                                                12m
remediation-service     ClusterIP   10.68.5.227    <none>        8080/TCP                                                12m
secret-service          ClusterIP   10.68.15.60    <none>        8080/TCP                                                12m
shipyard-controller     ClusterIP   10.68.10.66    <none>        8080/TCP                                                12m
statistics-service      ClusterIP   10.68.2.241    <none>        8080/TCP                                                12m
webhook-service         ClusterIP   10.68.5.152    <none>        8080/TCP                                                12m

```

**Reply by odubajDT** [reply]
Which steps did you follow during Keptn installation? I consider we can start from the beginning, so please type:

Install istio:
`curl -L https://istio.io/downloadIstio | ISTIO_VERSION=1.12.1 sh -`
`./istio-1.12.1/bin/istioctl install`

Then install Keptn:
`helm install keptn https://github.com/keptn/keptn/releases/download/0.12.0/keptn-0.12.0.tgz -n keptn --create-namespace --set=continuous-delivery.enabled=true --wait`
`helm install helm-service https://github.com/keptn/keptn/releases/download/0.12.0/helm-service-0.12.0.tgz -n keptn --create-namespace --wait`
`helm install jmeter-service https://github.com/keptn/keptn/releases/download/0.12.0/jmeter-service-0.12.0.tgz -n keptn --create-namespace --wait`

After these commands succeed, configure istio:
`curl -o configure-istio.sh https://raw.githubusercontent.com/keptn/examples/0.11.0/istio-configuration/configure-istio.sh`
`chmod +x configure-istio.sh`
`./configure-istio.sh`

And now authenticate:
`keptn auth --endpoint=$(kubectl -n keptn get ingress api-keptn-ingress -ojsonpath='{.spec.rules[0].host}')/api --api-token=$(kubectl get secret keptn-api-token -n keptn -ojsonpath='{.data.keptn-api-token}' | base64 --decode)`

Hope this will help you solve your problems. More info to the steps here: https://tutorials.keptn.sh/tutorials/keptn-full-tour-dynatrace-011/index.html?index=..%2F..index#0

If you have any other problems, just reach out.

**Reply by afzal442** [reply]
that's nice. but I'm trying to access Keptn via a **Port-forward**
Option 4: Access Keptn via a Port-forward
https://keptn.sh/docs/0.12.x/operate/install/#option-4-access-keptn-via-a-port-forward
`keptn auth --endpoint=$KEPTN_ENDPOINT --api-token=$KEPTN_API_TOKEN`
https://keptn.sh/docs/0.12.x/operate/install/#authenticate-keptn-cli
After running `keptn auth --endpoint=$KEPTN_ENDPOINT --api-token=$KEPTN_API_TOKEN`
Before that I ran `export KEPTN_ENDPOINT=http://localhost:8080/api` and `export  KEPTN_API_TOKEN=4W67hWnUoQQUab3LzAQBQPH4wVHNVnU8YKj6KBRHDinpo`
```
Kube context has been changed to gke_xyz_asia-south1-a_cluster-2Starting to authenticate
Authentication was unsuccessful. invalid character '<' looking for beginning of value
-----DETAILS-----<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx</center>
</body>
</html>
```

**Reply by afzal442** [reply]
Any clue @odubajDT ?

**Reply by odubajDT** [reply]
maybe trying to use LoadBalancer instead of ClusterIP for the gateway? I am just guessing where the problem might lie...

`kubectl patch svc api-gateway-nginx -p '{"spec": {"type": "LoadBalancer"}}' -n keptn`

**Reply by afzal442** [reply]
@odubajDT I tried that out. But failed. facing the same err again
`kubectl -n keptn get service api-gateway-nginx`
```
NAME                    TYPE           CLUSTER-IP     EXTERNAL-IP     PORT(S)                                                 AGE
api-gateway-nginx       LoadBalancer   10.68.1.160    34.93.113.189   80:32211/TCP                                            135m
```

**Reply by afzal442** [reply]
sorry I just figured it out that my cluster failed to schedule those resources. I'm re configuring that now.

**Reply by ghost** [reply]
Do you need any further assistance @afzal442 ?

**Reply by afzal442** [reply]
@Dynatrace-Adam-Gardner currently I'm having issue with deployment to gke cluster. I'm not able to provision `keptn deployment` resources due to size limit. Thanks.

**Reply by afzal442** [reply]
now I made it though. Could you tell me what the CLI command to trigger keptn delivery is? I tried `keptn trigger delivery --project=sockshop --service=carts --stage=dev --image=nginx --tag=1.14`
but failed
```
Availability of provided image cannot be checked.
Kube context has been changed to gke_xyz_us-central1-c_mycluster2
Starting to deliver the service carts in project sockshop in version nginx:1.14
Error: error while retrieving information for project sockshop: Project not found: sockshop
```
// @odubajDT

**Reply by ghost** [reply]
That warning suggests you do not have a `sockshop` project. Did you recreate it after you recreated your cluster. Please run `keptn get projects` and provide the output.

Your profile suggests you're based in India. I'm in Brisbane Australia, so perhaps we can arrange a call to walk through this today? Please email me (address on profile) and I'll set something up.

**Reply by afzal442** [reply]
thanks @Dynatrace-Adam-Gardner. That's fine.
I emailed you with my tz.
